quaternions.scad

```diff
@@ -214,65 +317,82 @@ function Q_Dist(q1, q2) = norm(q2-q1);
+    assert(Q_is_quat(q1) && Q_is_quat(q2), "Invalid quaternion(s)" )
     let(
-        dot = constrain(dot,-1,1),
-        theta_0 = acos(dot),
-        theta = theta_0 * u,
-        q3 = Q_Normalize(q2 - q1*dot),
-        out = q1*cos(theta) + q3*sin(theta)
-    ) out;
+      _dot  = q1*q2,
+      q1   = q1/norm(q1),
+      q2   = _dot<0 ? -q2/norm(q2) : q2/norm(q2),
+      dot = abs(_dot)
+    )
+    ! is_num(u) 
+    ? [for (uu=u) Q_Slerp(q1, q2, uu, dot)]
+    : Q_Slerp(q1, q2, u, dot) 
+  : _dot>0.9995
+    ? _Qnorm(q1 + u*(q2-q1))
+    : let( 
+        theta  = u*acos(_dot),
+        q3    = _Qnorm(q2 - _dot*q1)
+      ) 
+      _Qnorm(q1*cos(theta) + q3*sin(theta));
 
 
 // Function: Q_Matrix3()
 // Usage:
 //   Q_Matrix3(q);
 // Description:
 //   Returns the 3x3 rotation matrix for the given normalized quaternion q.
-function Q_Matrix3(q) = [
+function Q_Matrix3(q) =   
+  let( q = Q_Normalize(q) )
+  [
     [1-2*q[1]*q[1]-2*q[2]*q[2],   2*q[0]*q[1]-2*q[2]*q[3],   2*q[0]*q[2]+2*q[1]*q[3]],
     [  2*q[0]*q[1]+2*q[2]*q[3], 1-2*q[0]*q[0]-2*q[2]*q[2],   2*q[1]*q[2]-2*q[0]*q[3]],
     [  2*q[0]*q[2]-2*q[1]*q[3],   2*q[1]*q[2]+2*q[0]*q[3], 1-2*q[0]*q[0]-2*q[1]*q[1]]
-];
+  ];
 
 
 // Function: Q_Matrix4()
 // Usage:
 //   Q_Matrix4(q);
 // Description:
 //   Returns the 4x4 rotation matrix for the given normalized quaternion q.
-function Q_Matrix4(q) = [
+function Q_Matrix4(q) =    
+  let( q = Q_Normalize(q) )
+  [
     [1-2*q[1]*q[1]-2*q[2]*q[2],   2*q[0]*q[1]-2*q[2]*q[3],   2*q[0]*q[2]+2*q[1]*q[3], 0],
     [  2*q[0]*q[1]+2*q[2]*q[3], 1-2*q[0]*q[0]-2*q[2]*q[2],   2*q[1]*q[2]-2*q[0]*q[3], 0],
     [  2*q[0]*q[2]-2*q[1]*q[3],   2*q[1]*q[2]+2*q[0]*q[3], 1-2*q[0]*q[0]-2*q[1]*q[1], 0],
     [                        0,                         0,                         0, 1]
-];
+  ];
 
 
 // Function: Q_Axis()
 // Usage:
 //   Q_Axis(q)
 // Description:
 //   Returns the axis of rotation of a normalized quaternion `q`.
-function Q_Axis(q) = let(d = sqrt(1-(q[3]*q[3]))) (d==0)? [0,0,1] : [q[0]/d, q[1]/d, q[2]/d];
-
+//   The input doesn't need to be normalized.
+function Q_Axis(q) =   
+  assert( Q_is_quat(q) , "Invalid quaternion" )
+  let( d = norm(_Qvec(q)) )
+  approx(d,0)? [0,0,1] : _Qvec(q)/d;
 
 // Function: Q_Angle()
 // Usage:
-//   Q_Angle(q)
+//   a = Q_Angle(q)
+//   a12 = Q_Angle(q1,q2);
 // Description:
-// Returns the angle of rotation (in degrees) of a normalized quaternion `q`.
-function Q_Angle(q) = 2 * acos(q[3]);
+//   If only q1 is given, returns the angle of rotation (in degrees) of that quaternion.
+//   If both q1 and q2 are given, returns the angle (in degrees) between them.
+//   The input quaternions don't need to be normalized.
+function Q_Angle(q1,q2) =
+  assert(Q_is_quat(q1) && (is_undef(q2) || Q_is_quat(q2)), "Invalid quaternion(s)" )
+  let( n1 = is_undef(q2)? norm(_Qvec(q1)): norm(q1) )
+  is_undef(q2)
+  ? 2 * atan2(n1,_Qreal(q1))
+  : 2 * acos(q1*q2/n1/norm(q2)) ; 

```

Comment:
Good extension of functionality.
The `acos()` really should have it's input `constrain()`ed to -1 to 1, though, because FP rounding errors WILL cause problems otherwise.

Reply:
You are right, here `constrain()` should be used.